# Exercicis de programació numèrica.

**Objectius**
- Creació d'arrays amb la llibreria numpy.
- Creació de matrius amb la llibreria numpy.
- Ús dels mètodes de la llibreria numpy.
- Utilització de funcions.



In [1]:
import numpy as np
import pandas as pd

print("Setup Complete")

Setup Complete


### Exercici 1
Crea una funció que donat un Array d’una dimensió, et faci un resum estadístic bàsic de les dades. Si detecta que l’array té més d’una dimensió, ha de mostrar un missatge d’error.

In [2]:
prime_num = np.array([2, 3, 5, 7, 11, 13])
second_array = np.array([[2, 3, 5, 7, 11, 13],[2, 3, 5, 7, 11, 13]])

In [3]:
def array_dimension(array):
    if np.ndim(array)==1:
        n_array=pd.Series(array)
        basic_stats=n_array.describe()
        return basic_stats
    else:
        return "Use a one-dimensional array, please."

In [4]:
array_dimension(prime_num)

count     6.000000
mean      6.833333
std       4.400758
min       2.000000
25%       3.500000
50%       6.000000
75%      10.000000
max      13.000000
dtype: float64

In [5]:
array_dimension(second_array)

'Use a one-dimensional array, please.'

### Exercici 2
Crea una funció que et generi un quadrat NxN de nombres aleatoris entre el 0 i el 100.

In [6]:
def square_matrix(rows,columns):
    rng = np.random.default_rng(seed=1)
    return rng.integers(100, size = (rows,columns))

In [7]:
square_matrix(7,7)

array([[47, 51, 75, 95,  3, 14, 82],
       [94, 24, 31, 86, 42, 27, 82],
       [25, 40, 64, 54,  8,  2, 86],
       [75, 83, 53, 81, 32, 45, 78],
       [12, 30, 12, 45, 97, 13, 38],
       [40, 90, 20, 50, 26,  1, 75],
       [ 6, 28, 49, 48, 11, 98, 74]], dtype=int64)

### Exercici 3
Crea una funció que donada una taula de dues dimensions (NxM), et calculi els totals per fila i els totals per columna.

In [8]:
def aggregates_rows_cols(matrix):
    #aggregate across rows
    aggregate_rows = matrix.sum(axis=0)
    
    #aggregate across columns
    aggregate_cols = matrix.sum(axis=1)
    
    return """Agregate rows: {0},
    Agregate columns: {1}""".format(aggregate_rows, aggregate_cols)

In [9]:
data = square_matrix(3,4)
data

array([[47, 51, 75, 95],
       [ 3, 14, 82, 94],
       [24, 31, 86, 42]], dtype=int64)

In [10]:
aggregates_rows_cols(data)

'Agregate rows: [ 74  96 243 231],\n    Agregate columns: [268 193 183]'

### Exercici 4
Implementa manualment una funció que calculi el coeficient de correlació. Informa-te'n sobre els seus usos i interpretació.

We can find a trend line between two variables. The closer the data are to the line, the relationship between the two variables is relatively strong. We can use this to make predictions and inferences, aka **educated guesses**. We can quantify the strength of the relationship with **correlation**.

We should have low confidence in correlations made with a small datasets... 

When we're talking about correlation we're only talking about using straight lines. For correlation, a *p-value* tels us the probability that randomly drawn dots will result in a similarly strong relationship, or stonger. Thus, the smaller the **p-value**, the more confidence we have in the predictions we make with the line. 

- The **maximum value** for **correlation**, **1**, occurs whenever you can draw a **straight line** with a **positive slope** that goes through all the data. 
- And our confidence in how useful the relationship is depends on how much data we have. 
- Our confidence in our inferences depends on the amount of data we have collected and the **p-value**. We should have the most confidence in a relationship that is supported by the most data and has the smallest **p-value**.
- In other words, just because you have a lot of data and you have a lot of confidence in your guess if the **correlation value** is small, your guess will still be bad. 
- When a **straight line** with a **negative slope** can go through the center of every data point, then the **correlation** = **-1**.
- In both cases, if a **straight line** cannot go through all of the data, then we will get **correlation values** closer to **0**...

In [11]:
height = [60, 62, 65, 68, 70, 74]
weight=[140, 138, 150, 166, 190, 250]

In [12]:
def coef_corr(x,y):
    # import scipy
    from scipy import stats
    
    # r, p stands for correlation, p-value
    # see tupple unpacking
    r, p = stats.pearsonr(x, y)
    
    #round to seven decimals
    return """Correlation: {0}
    p-value: {1}""".format(round(r,7), round(p,7))

In [13]:
coef_corr(height, weight)

'Correlation: 0.9298975\n    p-value: 0.0071993'

The correlation between height and weight is **0.9298975**.

We have **high correlation**. This means that is extremely likely to have high correlation next time we collect six more samples.

We can also calculate correlation using *numpy*.

In [14]:
np.corrcoef(height, weight)

array([[1.        , 0.92989745],
       [0.92989745, 1.        ]])

In [15]:
np.corrcoef(height, weight)[0][1]

0.9298974521408027